xl-v1-multipart.go

```diff
@@ -442,25 +440,44 @@ func (xl xlObjects) PutObjectPart(bucket, object, uploadID string, partID int, s
 		return "", toObjectErr(errXLWriteQuorum, bucket, object)
 	}
 
-	updatedEInfos := make([]erasureInfo, len(onlineDisks))
+	var updatedEInfos []erasureInfo
 	for index := range partsMetadata {
-		updatedEInfos[index] = partsMetadata[index].Erasure
+		updatedEInfos = append(updatedEInfos, partsMetadata[index].Erasure)
 	}
 
+	var checksums []checkSumInfo
 	for index, eInfo := range newEInfos {
 		if eInfo.IsValid() {
-			updatedEInfos[index] = eInfo
-			for j := range updatedEInfos[index].Checksum {
-				updatedEInfos[index].Checksum[j] = eInfo.Checksum[j]
+
+			// Use a map to find union of checksums of parts that
+			// we concurrently written and committed before this
+			// part. N B For a different, concurrent upload of the
+			// same part, the last written content remains.
+			checksumSet := make(map[string]checkSumInfo)
+			checksums = updatedEInfos[index].Checksum
+			for _, cksum := range checksums {
+				checksumSet[cksum.Name] = cksum
 			}
+			checksums = newEInfos[index].Checksum
+			for _, cksum := range checksums {
+				checksumSet[cksum.Name] = cksum
+			}
+			// Form the checksumInfo to be committed in xl.json
+			// from the map.
+			var finalChecksums []checkSumInfo
+			for _, cksum := range checksumSet {
+				finalChecksums = append(finalChecksums, cksum)
+			}
+			updatedEInfos[index] = eInfo
+			updatedEInfos[index].Checksum = finalChecksums
```

Comment:
there is one corner case race condition:
1) upload part.1
2) upload part.2 and part.1 in parallel

there is a chance that part.1's old checksum is retained.